Framework/NexusGeometry/src/NexusGeometrySave.cpp

```diff
@@ -174,31 +215,182 @@ inline H5::Group simpleNXSubGroup(H5::Group &parent, const std::string &name,
  * @param compInfo : Component Info Instrument cache
  * @param idx : index of bank in cache.
  */
-inline void writeXYZPixeloffset(H5::Group &grp,
-                                const Geometry::ComponentInfo &compInfo,
-                                const size_t idx) {
+inline void writePixelData(H5::Group &grp,
+                           const Geometry::ComponentInfo &compInfo,
+                           const size_t idx) {
 
-  H5::DataSet xPixelOffset, yPixelOffset, zPixelOffset;
-  auto childrenDetectors = compInfo.detectorsInSubtree(idx);
+  // write pixel offset
+
+  H5::DataSet xPixelOffset, yPixelOffset, zPixelOffset; // pixel offset datasets
+  auto pixels =
+      compInfo.detectorsInSubtree(idx); // indices of all pixels in bank
+  const auto nPixels = pixels.size();   // number of pixels
+
+  // prevents undefined behaviour when passing empty container into std::all_of.
+  // If children detectors empty, there is no data to write to bank. exits here.
+  if (pixels.empty())
+    return;
+
+  bool shapesAreHomogeneous{false};         // all shapes equal
+  bool shapesAreValidAndHomogeneous{false}; // all shapes equal and valid shape
+
+  // shape type of the first detector in children detectors
+  auto &firstShape = compInfo.shape(pixels.front());
+  auto firstShapeInfo = firstShape.shapeInfo();
+  auto fType = firstShapeInfo.shape();
+  auto fHeight = firstShapeInfo.height();
+  auto fRadius = firstShapeInfo.radius();
+
+  // return true if all shape types, heights and radii are equal to the first
+  // shape in children detectors.
+  shapesAreHomogeneous =
+      std::all_of(pixels.begin(), pixels.end(), [&](const size_t &idx) {
+        auto &shapeObj = compInfo.shape(idx);
+
+        if (auto *meshObject =
+                dynamic_cast<const Geometry::MeshObject *>(&firstShape)) {
+          // current implementation only considers solid shapes
+          return false;
+        }
+        if (auto *meshObject2d =
+                dynamic_cast<const Geometry::MeshObject2D *>(&firstShape)) {
+          // current implementation only considers solid shapes
+          return false;
+        }
+        auto shapeInfo = shapeObj.shapeInfo();
```

Comment:
you don't check that the shape info corresponds to a cylinder. This is an important check since we do NOT handle any other kind of CSGObject